README
AutomysqlBackup
-------------------------
.. INDEX
-------------------------
Disclaimer
Install
Usage
Configuration Options
Encryption
Backup rotation
Restoring



.. DISCLAIMER
-------------------------
I take no resposibility for any data loss or corruption when using this script.
This script will not help in the event of a hard drive crash. If a copy of the
backup has not been stored offline or on another PC. You should copy your backups
offline regularly for best protection.

Happy backing up...



.. INSTALL
-------------------------
Extract the package into a directory. If you are reading this you have probably done
this already.

To install the Automysqlbackup the easy way.
1. Run the install.sh script.
2. Edit the /etc/automysqlbackup/myserver.conf file to customise your settings.
3. See usage section.

To install it manually (the hard way).
1. Create the /etc/automysqlbackup directory.
2. Copy in the automysqlbackup.conf file.
3. Copy the automysqlbackup file to /usr/local/bin and make executable.
4. cp /etc/automysqlbackup/automysqlbackup.conf /etc/automysqlbackup/myserver.conf
5. Edit the /etc/automysqlbackup/myserver.conf file to customise your settings.
6. See usage section.



.. USAGE
-------------------------

Automysqlbackup can be run a number of ways, you can choose which is best for you.

1. Create a script as below called runmysqlbackup using the lines below:

#~~~~ Copy From Below Here ~~~~
#!/bin/sh

/usr/local/bin/automysqlbackup /etc/automysqlbackup/myserver.conf

chown root.root /var/backup/db* -R
find /var/backup/db* -type f -exec chmod 400 {} \;
find /var/backup/db* -type d -exec chmod 700 {} \;

#~~~~~ Copy To Above Here ~~~~

2. Save it to a suitable location or copy it to your /etc/cron.daily folder. 

3. Make it executable, i.e. chmod +x /etc/cron.daily/runmysqlbackup.


The backup can be run from the command line simply by running the following command.

  automysqlbackup /etc/automysqlbackup/myserver.conf

If you don't supply an argument for automysqlbackup, the default configuration
in the program automysqlbackup will be used unless a global file

  CONFIG_configfile="/etc/automysqlbackup/automysqlbackup.conf"

exists.

You can just copy the supplied automysqlbackup.conf as many times as you want
and use for separate configurations, i.e. for example different mysql servers.

!!! NEW !!!
----------
As of version 3.0 we have added differential backups using the program diff. In an
effort to make the reconstruction of the full archives more user friendly, we
added new functionality to the script. Therefore, while preserving the old syntax,
we created options for the script, so that the new functions can be accessed.

Usage automysqlbackup options -cblh
-c CONFIG_FILE  Specify optional config file.
-b      Use backup method.
-l      List manifest entries.
-h      Show this help.

If you use these options, you have to specify everything according to them and can't
mix the old syntax with the new one. Example:

before (still valid!):

  >> automysqlbackup "myconfig.conf"

now:

  >> automysqlbackup -c "myconfig.conf" -b

which is equivalent to

  >> automysqlbackup -bc "myconfig.conf"

or in English: The order of the options doesn't matter, however those options expecting
arguments, have to be placed right before the argument (as seen above).

The option '-l' (List manifest entries) finds all Manifest files in your configuration
directory (you need to specify your optional config file - otherwise a fallback will be
used: global config file -> program internal default options). It then filters from which
databases these are and presents you with a list (you can select more than one!) of them.
Once you have chosen, you will be given a list of Manifest files, from which you choose
again and after that from which you choose differential files. When you have completed
all your selections, a list of selected differential files will be shown. You may then
choose what you want to be done with/to those files. At the moment the options are:
- create full backup out of differential one
- remove the differential backup and its Manifest entry.


.. CONFIGURATION OPTIONS
-------------------------

!! "automysqlbackup" program contains a default configuration that should not be changed:

The global config file which overwrites the default configuration is located here
"/etc/automysqlbackup/automysqlbackup.conf" by default.

Please take a look at the supplied "automysqlbackup.conf" for information about the configuration options.

Default configuration
CONFIG_configfile="/etc/automysqlbackup/automysqlbackup.conf"
CONFIG_backup_dir='/var/backup/db'
CONFIG_do_monthly="01"
CONFIG_do_weekly="5"
CONFIG_rotation_daily=6
CONFIG_rotation_weekly=35
CONFIG_rotation_monthly=150
CONFIG_mysql_dump_usessl='yes'
CONFIG_mysql_dump_username='root'
CONFIG_mysql_dump_password=''
CONFIG_mysql_dump_host='localhost'
CONFIG_mysql_dump_socket=''
CONFIG_mysql_dump_create_database='no'
CONFIG_mysql_dump_use_separate_dirs='yes'
CONFIG_mysql_dump_compression='gzip'
CONFIG_mysql_dump_commcomp='no'
CONFIG_mysql_dump_latest='no'
CONFIG_mysql_dump_max_allowed_packet=''
CONFIG_db_names=()
CONFIG_db_month_names=()
CONFIG_db_exclude=( 'information_schema' )
CONFIG_mailcontent='log'
CONFIG_mail_maxattsize=4000
CONFIG_mail_address='root'
CONFIG_encrypt='no'
CONFIG_encrypt_password='password0123'

!! automysqlbackup (the shell program) accepts one parameter, the filename of a configuration file. The entries in there will supersede all others.

Please take a look at the supplied "automysqlbackup.conf" for information about the configuration options.



.. ENCRYPTION
-------------------------

To decrypt run (replace bz2 with gz if using gzip):

openssl enc -aes-256-cbc -d -in encrypted_file_name(ex: *.enc.bz2) -out outputfilename.bz2 -pass pass:PASSWORD-USED-TO-ENCRYPT



.. BACKUP ROTATION
-------------------------

Daily Backups are rotated weekly.
Weekly Backups are run on fridays, unless otherwise specified via CONFIG_do_weekly.
Weekly Backups are rotated on a 5 week cycle, unless otherwise specified via CONFIG_rotation_weekly.
Monthly Backups are run on the 1st of the month, unless otherwise specified via CONFIG_do_monthly.
Monthly Backups are rotated on a 5 month cycle, unless otherwise specified via CONFIG_rotation_monthly.

Suggestion: It may be a good idea to copy monthly backups offline or to another server.



.. RESTORING
-------------------------

Firstly you will need to uncompress the backup file and decrypt it if encryption was used (see encryption section).

eg.
gunzip file.gz (or bunzip2 file.bz2)

Next you will need to use the mysql client to restore the DB from the sql file.

eg.
  mysql --user=username --pass=password --host=dbserver database < /path/file.sql
or
  mysql --user=username --pass=password --host=dbserver -e "source /path/file.sql" database

NOTE: Make sure you use "<" and not ">" in the above command because you are piping the file.sql to mysql and not the other way around.



CHANGELOG
#=====================================================================
# Change Log
#=====================================================================
#
# version 3.0.7 - (2018-05-02)
#     - Small changes for MacOS
# version 3.0.6 - (2018-01-10)
#     - Fix the mysql login-configuration file cannot be located in a cronjob
# version 3.0.5 - (2017-12-08)
#     - Changed usessl to no per default
# version 3.0.4 - (2017-10-27)
#     - Adjusted md5sums in install.sh
# version 3.0.3 - (2017-09-27)
#     - Configuration fix
# version 3.0.2 - (2017-01-20)
#     - Add secure connection config support to install.sh
# version 3.0.1 - (2016-10-05)
#     - Forked from Sourceforge
#     - Supports MySQL 5.6 and MySQL 5.7
#     - Updated README for login path support
# version 3.0_rc4 - (2011-11-24)
#     - Removing mkfifo commands, thereby improving portability.
#     - Fixing not working table exclude feature for wildcards.
# version 3.0_rc3 - (2011-11-24)
#     - Changed code to make it more portable, thereby resolving FreeBSD issue.
# version 3.0_rc2 - (2011-11-21)
#     - Added multicore support for bzip and gzip2.
#     - Fixed error in README file.
# version 3.0_rc1 - (2011-11-15)
#     - Added differential backup method.
#     - Added user-friendly method to recreate full backups out of differential ones.
#     - Changed paramters, with which the script can be called, to make the new
#       methods available.
#     - Fixed some bugs.
# version 3.0_beta2 - (2011-08-21)
#     - Added possibility to backup local files. 
#     - Added full schema backup.
#     - Added master-data option.
#     - Fixed some bugs.
# version 3.0_beta1 - (2011-08-15)
#     - REMOVED: Implementation of Variables containing full path to binaries to
#       avoid possibly confusion with aliases or builtins. (by Johannes Kolter)
#       Inside a bash script no aliases are used! This didn't make ANY sense!
#       Thereby resolved bug item #3074425.
#     - Changed some variables to be arrays, i.e. lists, and removed the ugly
#       sed stuff.
#       Fixed bug item #3169562
#     - Added check for weekly and monthly backups, so that they are unique per day.
#       bug item #3185389
#     - Changed SHEBANG to #!/usr/bin/env bash  for portability reasons.
#       bug item #3292873
#     - Changed config file structure: read /etc/automysqlbackup/mysqlbackup.conf,
#       if supplied read configfile parameter (no more -c or sth. like that, just
#       the name of the file!),
#       whatever isn't set yet, set in here to default values.
#     - bug item #3110715: create a file in /etc/cron.d/daily and call the script
#       from in there, i.e. place the script in /usr/local/bin
#     - bug item #3082899: the PATH variables are different in ssh, you have to
#       supply the complete path
#     - Fixed bug item #3064547, suggestion accepted.
#     - Fixed bug item #3031023, suggestion accepted.
#     - Fixed bug item #3030604, resolved due to design correction.
#     - Fixed bug item #3025849, as long as basename is in $PATH on your system.
#     - Fixed bug item #3030478.
#     - bug item #3054633: .muttrc entry save=yes will result in saving sent files!
#     - Feature request item #1538588.
#     - Feature request item #1538138.
#     - Feature request item #1538142.
#     - Feature request item #1541843 was already included.
#     - Feature request item #2808012.
#     - Feature request item #2831465.
#     - Feature request item #3052484. Mysqldump already has an ssl option.
#     - Feature request item #3190079. I hope cleaning up everything older than 24
#       hours as a lower limit is good enough.
#     - Feature request item #3284779 was already included. See CONFIG_mysql_dump_latest.
#     - Feature request item #3053623.
# version 2.6.0 - (2011-07-19)
#     - Fixed bug where files would not email correctly (Fix by Jesse Vaughan)
#     - Added section to encrypt .gz and .bz2 files using openssl (added by Jesse Vaughan)
# version 2.5.1-01 - (2010-07-06)
#     - Fixed pathname bug item #3025849 (by Johannes Kolter)
# version 2.5.1 - (2010-07-04)
#     - Added support for default and optional config file (by Johannes Kolter)
#     - Rotating after backup was successful whith find(1) (by Johannes Kolter)
#     - Implementation of Variables containing full path to binaries to
#       avoid possibly confusion with aliases or builtins. (by Johannes Kolter)
#     - Fixed bug where weekly backups were not being rotated.
#       Added rotation of 5 monthly backups
#       Now all old backups are deleted, not only the most recent one
#       (inspired by oleg@bintime.com)
#     - Use Debian special-file to access database (by Johannes Kolter)
#     - Fixed bug ID: 1438565
#       Moved IO redirection to a place before decicions are made and actions are taken.
#       (inspired by Derk Bernhardt)
#     - Fixed bug ID: #3000316 (reported by Sascha Feldhorst)
#     - Fixed bug ID: #1529458 (reported by Natalie ( njwood ))
#     - Fixed bug ID: #1548919 (reported by Piotr Kuczynski)
# version 2.5 - (2006-01-15)

install.sh
#!/usr/bin/env bash
#

# return	true, if variable is set; else false
isSet() {
  if [[ ! ${!1} && ${!1-_} ]]; then return 1; else return 0; fi
}

activateIO() {
    touch "$1"
    exec 6>&1
    exec > "$1"
}
removeIO() {
  exec 1>&6 6>&-
}

upgrade_config_file () {
( # execute in subshell, so that sourced variables are only available inside ()
    source "$1"
    #declare -p
    local temp
    temp=$(mktemp /tmp/tmp.XXXXXX)
    (( $? != 0 )) && return 1
    activateIO "$temp"
    echo "#version=3.0_rc2"
    echo "# Uncomment to change the default values (shown after =)"
    echo "# WARNING:"
    echo "# This is not true for UMASK, CONFIG_prebackup and CONFIG_postbackup!!!"
    echo "#"
    echo "# Default values are stored in the script itself. Declarations in"
    echo "# /etc/automysqlbackup/automysqlbackup.conf will overwrite them. The"
    echo "# declarations in here will supersede all other."
    echo ""
    echo "# Edit \$PATH if mysql and mysqldump are not located in /usr/local/bin:/usr/bin:/bin:/usr/local/mysql/bin"
    echo "#PATH=\${PATH}:FULL_PATH_TO_YOUR_DIR_CONTAINING_MYSQL:FULL_PATH_TO_YOUR_DIR_CONTAINING_MYSQLDUMP"
    echo ""
    echo "# Basic Settings"
    echo ""
	echo "# since mysql 5.6.x connections can be stored securely"
	echo "# add your connection with"
	echo "# mysql_config_editor set --login-path=automysqldump --host=localhost --user=root --password"
	echo "# automysqldump is using the login-path "automysqldump" as default"
    echo ""
    echo "# Username to access the MySQL server e.g. dbuser"
    if isSet USERNAME; then
      printf '%s=%q\n' CONFIG_mysql_dump_username "${USERNAME-}"
    else
      echo "#CONFIG_mysql_dump_username='root'"
    fi
    echo ""
    echo "# Password to access the MySQL server e.g. password"
    if isSet PASSWORD; then
      printf '%s=%q\n' CONFIG_mysql_dump_password "${PASSWORD-}"
    else
      echo "#CONFIG_mysql_dump_password=''"
    fi
    echo ""
    echo "# Host name (or IP address) of MySQL server e.g localhost"
    if isSet DBHOST; then
      printf '%s=%q\n' CONFIG_mysql_dump_host "${DBHOST-}"
    else
      echo "#CONFIG_mysql_dump_host='localhost'"
    fi
    echo ""
    echo "# \"Friendly\" host name of MySQL server to be used in email log"
    echo "# if unset or empty (default) will use CONFIG_mysql_dump_host instead"
    if isSet CONFIG_mysql_dump_host_friendly; then
      printf '%s=%q\n' CONFIG_mysql_dump_host_friendly "${CONFIG_mysql_dump_host_friendly-}"
    else
      echo "#CONFIG_mysql_dump_host_friendly=''"
    fi
    echo ""
    echo "# Backup directory location e.g /backups"
    if isSet BACKUPDIR; then
      printf '%s=%q\n' CONFIG_backup_dir "${BACKUPDIR-}"
    else
      echo "#CONFIG_backup_dir='/var/backup/db'"
    fi
    echo ""
    echo "# This is practically a moot point, since there is a fallback to the compression"
    echo "# functions without multicore support in the case that the multicore versions aren't"
    echo "# present in the system. Of course, if you have the latter installed, but don't want"
    echo "# to use them, just choose no here."
    echo "# pigz -> gzip"
    echo "# pbzip2 -> bzip2"
    echo "#CONFIG_multicore='yes'"
    echo ""
    echo "# Number of threads (= occupied cores) you want to use. You should - for the sake"
    echo "# of the stability of your system - not choose more than (#number of cores - 1)."
    echo "# Especially if the script is run in background by cron and the rest of your system"
    echo "# has already heavy load, setting this too high, might crash your system. Assuming"
    echo "# all systems have at least some sort of HyperThreading, the default is 2 threads."
    echo "# If you wish to let pigz and pbzip2 autodetect or use their standards, set it to"
    echo "# 'auto'."
    echo "#CONFIG_multicore_threads=2"
    echo ""
    echo "# Databases to backup"
    echo ""
    echo "# List of databases for Daily/Weekly Backup e.g. ( 'DB1' 'DB2' 'DB3' ... )"
    echo "# set to (), i.e. empty, if you want to backup all databases"
    if isSet DBNAMES; then
      if [[ "x$DBNAMES" = "xall" ]]; then
	echo "#CONFIG_db_names=()"
      else
	declare -a CONFIG_db_names
	for i in $DBNAMES; do
	  CONFIG_db_names=( "${CONFIG_db_names[@]}" "$i" )
	done
	declare -p CONFIG_db_names | sed -e 's/\[[^]]*]=//g'
      fi
    else
      echo "#CONFIG_db_names=()"
    fi
    echo "# You can use"
    echo "#declare -a MDBNAMES=( \"\${DBNAMES[@]}\" 'added entry1' 'added entry2' ... )"
    echo "# INSTEAD to copy the contents of \$DBNAMES and add further entries (optional)."
    echo ""
    echo "# List of databases for Monthly Backups."
    echo "# set to (), i.e. empty, if you want to backup all databases"
    if isSet MDBNAMES; then
      if [[ "x$MDBNAMES" = "xall" ]]; then
	echo "#CONFIG_db_month_names=()"
      else
	declare -a CONFIG_db_month_names
	for i in $MDBNAMES; do
	  CONFIG_db_month_names=( "${CONFIG_db_month_names[@]}" "$i" )
	done
	declare -p CONFIG_db_month_names | sed -e 's/\[[^]]*]=//g'
      fi
    else
      echo "#CONFIG_db_month_names=()"
    fi
    echo ""
    echo "# List of DBNAMES to EXLUCDE if DBNAMES is empty, i.e. ()."
    if isSet DBEXCLUDE; then
      declare -a CONFIG_db_exclude
      for i in $DBEXCLUDE; do
	CONFIG_db_exclude=( "${CONFIG_db_exclude[@]}" "$i" )
      done
      declare -p CONFIG_db_exclude | sed -e 's/\[[^]]*]=//g'
    else
      echo "#CONFIG_db_exclude=( 'information_schema' )"
    fi
    echo ""
    echo "# List of tables to exclude, in the form db_name.table_name"
    echo "#CONFIG_table_exclude=()"
    echo ""
    echo ""
    echo "# Advanced Settings"
    echo ""
    echo "# Rotation Settings"
    echo ""
    echo "# Which day do you want monthly backups? (01 to 31)"
    echo "# If the chosen day is greater than the last day of the month, it will be done"
    echo "# on the last day of the month."
    echo "# Set to 0 to disable monthly backups."
    echo "#CONFIG_do_monthly=\"01\""
    echo ""
    echo "# Which day do you want weekly backups? (1 to 7 where 1 is Monday)"
    echo "# Set to 0 to disable weekly backups."
    if isSet DOWEEKLY; then
      printf '%s=%q\n' CONFIG_do_weekly "${DOWEEKLY-}"
    else
      echo "#CONFIG_do_weekly=\"5\""
    fi
    echo ""
    echo "# Set rotation of daily backups. VALUE*24hours"
    echo "# If you want to keep only today's backups, you could choose 1, i.e. everything older than 24hours will be removed."
    echo "#CONFIG_rotation_daily=6"
    echo ""
    echo "# Set rotation for weekly backups. VALUE*24hours"
    echo "#CONFIG_rotation_weekly=35"
    echo ""
    echo "# Set rotation for monthly backups. VALUE*24hours"
    echo "#CONFIG_rotation_monthly=150"
    echo ""
    echo ""
    echo "# Server Connection Settings"
    echo ""
    echo "# Set the port for the mysql connection"
    echo "#CONFIG_mysql_dump_port=3306"
    echo ""
    echo "# Compress communications between backup server and MySQL server?"
    if isSet COMMCOMP; then
      printf '%s=%q\n' CONFIG_mysql_dump_commcomp "${COMMCOMP-}"
    else
      echo "#CONFIG_mysql_dump_commcomp='no'"
    fi
    echo ""
    echo "# Use ssl encryption with mysqldump?"
    echo "#CONFIG_mysql_dump_usessl='yes'"
    echo ""
    echo "# For connections to localhost. Sometimes the Unix socket file must be specified."
    if isSet SOCKET; then
      printf '%s=%q\n' CONFIG_mysql_dump_socket "${SOCKET-}"
    else
      echo "#CONFIG_mysql_dump_socket=''"
    fi
    echo ""
    echo "# The maximum size of the buffer for client/server communication. e.g. 16MB (maximum is 1GB)"
    if isSet MAX_ALLOWED_PACKET; then
      printf '%s=%q\n' CONFIG_mysql_dump_max_allowed_packet "${MAX_ALLOWED_PACKET-}"
    else
      echo "#CONFIG_mysql_dump_max_allowed_packet=''"
    fi
    echo ""
    echo "# This option sends a START TRANSACTION SQL statement to the server before dumping data. It is useful only with"
    echo "# transactional tables such as InnoDB, because then it dumps the consistent state of the database at the time"
    echo "# when BEGIN was issued without blocking any applications."
    echo "#"
    echo "# When using this option, you should keep in mind that only InnoDB tables are dumped in a consistent state. For"
    echo "# example, any MyISAM or MEMORY tables dumped while using this option may still change state."
    echo "#"
    echo "# While a --single-transaction dump is in process, to ensure a valid dump file (correct table contents and"
    echo "# binary log coordinates), no other connection should use the following statements: ALTER TABLE, CREATE TABLE,"
    echo "# DROP TABLE, RENAME TABLE, TRUNCATE TABLE. A consistent read is not isolated from those statements, so use of"
    echo "# them on a table to be dumped can cause the SELECT that is performed by mysqldump to retrieve the table"
    echo "# contents to obtain incorrect contents or fail."
    echo "#CONFIG_mysql_dump_single_transaction='no'"
    echo ""
    echo "# http://dev.mysql.com/doc/refman/5.0/en/mysqldump.html#option_mysqldump_master-data"
    echo "# --master-data[=value] "
    echo "# Use this option to dump a master replication server to produce a dump file that can be used to set up another"
    echo "# server as a slave of the master. It causes the dump output to include a CHANGE MASTER TO statement that indicates"
    echo "# the binary log coordinates (file name and position) of the dumped server. These are the master server coordinates"
    echo "# from which the slave should start replicating after you load the dump file into the slave."
    echo "#"
    echo "# If the option value is 2, the CHANGE MASTER TO statement is written as an SQL comment, and thus is informative only;"
    echo "# it has no effect when the dump file is reloaded. If the option value is 1, the statement is not written as a comment"
    echo "# and takes effect when the dump file is reloaded. If no option value is specified, the default value is 1."
    echo "#"
    echo "# This option requires the RELOAD privilege and the binary log must be enabled. "
    echo "#"
    echo "# The --master-data option automatically turns off --lock-tables. It also turns on --lock-all-tables, unless"
    echo "# --single-transaction also is specified, in which case, a global read lock is acquired only for a short time at the"
    echo "# beginning of the dump (see the description for --single-transaction). In all cases, any action on logs happens at"
    echo "# the exact moment of the dump."
    echo "# =================================================================================================================="
    echo "# possible values are 1 and 2, which correspond with the values from mysqldump"
    echo "# VARIABLE=    , i.e. no value, turns it off (default)"
    echo "#"
    echo "#CONFIG_mysql_dump_master_data="
    echo ""
    echo "# Included stored routines (procedures and functions) for the dumped databases in the output. Use of this option"
    echo "# requires the SELECT privilege for the mysql.proc table. The output generated by using --routines contains"
    echo "# CREATE PROCEDURE and CREATE FUNCTION statements to re-create the routines. However, these statements do not"
    echo "# include attributes such as the routine creation and modification timestamps. This means that when the routines"
    echo "# are reloaded, they will be created with the timestamps equal to the reload time."
    echo "#"
    echo "# If you require routines to be re-created with their original timestamp attributes, do not use --routines. Instead,"
    echo "# dump and reload the contents of the mysql.proc table directly, using a MySQL account that has appropriate privileges"
    echo "# for the mysql database. "
    echo "#"
    echo "# This option was added in MySQL 5.0.13. Before that, stored routines are not dumped. Routine DEFINER values are not"
    echo "# dumped until MySQL 5.0.20. This means that before 5.0.20, when routines are reloaded, they will be created with the"
    echo "# definer set to the reloading user. If you require routines to be re-created with their original definer, dump and"
    echo "# load the contents of the mysql.proc table directly as described earlier."
    echo "#"
    echo "#CONFIG_mysql_dump_full_schema='yes'"
    echo ""
    echo "# Backup dump settings"
    echo ""
    echo "# Include CREATE DATABASE in backup?"
    if isSet CREATE_DATABASE; then
      printf '%s=%q\n' CONFIG_mysql_dump_create_database "${CREATE_DATABASE-}"
    else
      echo "#CONFIG_mysql_dump_create_database='no'"
    fi
    echo ""
    echo "# Separate backup directory and file for each DB? (yes or no)"
    if isSet SEPDIR; then
      printf '%s=%q\n' CONFIG_mysql_dump_use_separate_dirs "${SEPDIR-}"
    else
      echo "#CONFIG_mysql_dump_use_separate_dirs='yes'"
    fi
    echo ""
    echo "# Choose Compression type. (gzip or bzip2)"
    if isSet COMP; then
      printf '%s=%q\n' CONFIG_mysql_dump_compression "${COMP-}"
    else
      echo "#CONFIG_mysql_dump_compression='gzip'"
    fi
    echo ""
    echo "# Store an additional copy of the latest backup to a standard"
    echo "# location so it can be downloaded by third party scripts."
    if isSet LATEST; then
      printf '%s=%q\n' CONFIG_mysql_dump_latest "${LATEST-}"
    else
      echo "#CONFIG_mysql_dump_latest='no'"
    fi
    echo ""
    echo "# Remove all date and time information from the filenames in the latest folder."
    echo "# Runs, if activated, once after the backups are completed. Practically it just finds all files in the latest folder"
    echo "# and removes the date and time information from the filenames (if present)."
    echo "#CONFIG_mysql_dump_latest_clean_filenames='no'"
    echo ""
    echo "# Notification setup"
    echo ""
    echo "# What would you like to be mailed to you?"
    echo "# - log   : send only log file"
    echo "# - files : send log file and sql files as attachments (see docs)"
    echo "# - stdout : will simply output the log to the screen if run manually."
    echo "# - quiet : Only send logs if an error occurs to the MAILADDR."
    if isSet MAILCONTENT; then
      printf '%s=%q\n' CONFIG_mailcontent "${MAILCONTENT-}"
    else
      echo "#CONFIG_mailcontent='stdout'"
    fi
    echo ""
    echo "# Set the maximum allowed email size in k. (4000 = approx 5MB email [see docs])"
    if isSet MAXATTSIZE; then
      printf '%s=%q\n' CONFIG_mail_maxattsize "${MAXATTSIZE-}"
    else
      echo "#CONFIG_mail_maxattsize=4000"
    fi
    echo ""
    echo "# Email Address to send mail to? (user@domain.com)"
    if isSet MAILADDR; then
      printf '%s=%q\n' CONFIG_mail_address "${MAILADDR-}"
    else
      echo "#CONFIG_mail_address='root'"
    fi
    echo ""
	echo '# Create differential backups. Master backups are created weekly at #$CONFIG_do_weekly weekday. Between master backups,'
	echo "# diff is used to create differential backups relative to the latest master backup. In the Manifest file, you find the"
	echo "# following structure"
	echo '# $filename 	md5sum	$md5sum	diff_id	$diff_id	rel_id	$rel_id'
	echo "# where each field is separated by the tabular character '\t'. The entries with $ at the beginning mean the actual values,"
	echo "# while the others are just for readability. The diff_id is the id of the differential or master backup which is also in"
	echo "# the filename after the last _ and before the suffixes begin, i.e. .diff, .sql and extensions. It is used to relate"
	echo '# differential backups to master backups. The master backups have 0 as $rel_id and are thereby identifiable. Differential'
	echo '# backups have the id of the corresponding master backup as $rel_id.'
	echo "#"
	echo '# To ensure that master backups are kept long enough, the value of $CONFIG_rotation_daily is set to a minimum of 21 days.'
	echo "#"
	echo "#CONFIG_mysql_dump_differential='no'"
    echo ""
    echo "# Encryption"
    echo ""
    echo "# Do you wish to encrypt your backups using openssl?"
    echo "#CONFIG_encrypt='no'"
    echo ""
    echo "# Choose a password to encrypt the backups."
    echo "#CONFIG_encrypt_password='password0123'"
    echo ""
    echo "# Other"
    echo ""
    echo "# Backup local files, i.e. maybe you would like to backup your my.cnf (mysql server configuration), etc."
    echo "# These files will be tar'ed, depending on your compression option CONFIG_mysql_dump_compression compressed and"
    echo "# depending on the option CONFIG_encrypt encrypted."
    echo "#"
    echo "# Note: This could also have been accomplished with CONFIG_prebackup or CONFIG_postbackup."
    echo "#CONFIG_backup_local_files=()"
    echo ""
    echo "# Command to run before backups (uncomment to use)"
    if isSet PREBACKUP; then
      printf '%s=%q\n' CONFIG_prebackup "${PREBACKUP-}"
    else
      echo "#CONFIG_prebackup=\"/etc/mysql-backup-pre\""
    fi
    echo ""
    echo "# Command run after backups (uncomment to use)"
    if isSet POSTBACKUP; then
      printf '%s=%q\n' CONFIG_postbackup "${POSTBACKUP-}"
    else
      echo "#CONFIG_postbackup=\"/etc/mysql-backup-post\""
    fi
    echo ""
    echo "# Uncomment to activate! This will give folders rwx------"
    echo "# and files rw------- permissions."
    echo "#umask 0077"
    echo ""
    echo "# dry-run, i.e. show what you are gonna do without actually doing it"
    echo "# inactive: =0 or commented out"
    echo "# active: uncommented AND =1"
    echo "#CONFIG_dryrun=1"
    removeIO
    mv "$temp" "${1}_converted"
    return 0
  )
}


parse_config_file () {
  printf 'Found config file %s. ' "$1"
  if head -n1 "$1" | egrep -o 'version=.*' >& /dev/null; then
    version=`head -n1 "$1" | egrep -o 'version=.*' | awk -F"=" '{print $2}'`
    if [[ "$version" =~ 3.* ]]; then
      printf 'Version 3.* determined. No conversion necessary.\n'
    else
      printf 'Unknown version. Can not convert it. You have to convert it manually.\n'
    fi
  else
    printf 'No version information on first line of config file. Assuming the version is <3.\n'
    while true; do
	read -p "Convert? [Y/n] " yn
	[[ "x$yn" = "x" ]] && { upgrade_config_file "$1" || echo "Failed to convert."; break; }
	case $yn in
	    [Yy]* ) upgrade_config_file "$1" || echo "Failed to convert."; break;;
	    [Nn]* ) break;;
	    * ) echo "Please answer yes or no.";;
	esac
    done
  fi
}

#precheck
echo "### Checking archive files for existence, readability and integrity."
echo

printf 'Select the global configuration directory [/etc/automysqlbackup]: '
read configdir
configdir="${configdir%/}" # strip trailing slash if there
[[ "x$configdir" = "x" ]] && configdir='/etc/automysqlbackup'
printf 'Select directory for the executable [/usr/local/bin]: '
read bindir
bindir="${bindir%/}" # strip trailing slash if there
[[ "x$bindir" = "x" ]] && bindir='/usr/local/bin'

#create global config directory
echo "### Creating global configuration directory ${configdir}:"
echo
if [ -d "${configdir}" ]; then
  echo "exists already ... searching for config files:"
  for i in "${configdir}"/*.conf; do
    [[ "x$(basename $i)" = "xautomysqlbackup.conf" ]] && continue
    parse_config_file "$i"
  done
else
  if mkdir "${configdir}" >/dev/null 2>&1; then
    #testing for permissions
    if [ -r "${configdir}" -a -x "${configdir}" ]; then
      printf "success\n"
    else
      printf "directory successfully created but has wrong permissions, trying to correct ... "
      if chmod +rx "${configdir}" >/dev/null 2>&1; then
	printf "corrected\n"
      else
	printf "failed. Aborting. Make sure you run the script with appropriate permissions.\n"
      fi
    fi
  else
    printf "failed ... check permissions.\n"
  fi
fi

echo
#copying files
echo "### Copying files."
echo
cp -i automysqlbackup.conf LICENSE README "${configdir}"/
cp -i automysqlbackup.conf "${configdir}"/myserver.conf
cp -i automysqlbackup "${bindir}"/
[[ -f "${bindir}"/automysqlbackup ]] && [[ -x "${bindir}"/automysqlbackup ]] || chmod +x "${bindir}"/automysqlbackup || echo " failed - make sure you make the program executable, i.e. run 'chmod +x ${bindir}/automysqlbackup'"
echo

if echo $PATH | grep "${bindir}" >/dev/null 2>&1; then
  printf "if you are running automysqlbackup under the same user as you run this install script,\nyou should be able to access it by running 'automysqlbackup' from the command line.\n"
  printf "if not, you have to check if 'echo \$PATH' has ${bindir} in it\n"
else
  printf "if running under the current user, you have to use the full path ${bindir}/automysqlbackup since /usr/local/bin is not in 'echo \$PATH'\n"
fi
printf "\nSetup Complete!\n"
